api/src/main/java/org/eclipse/microprofile/reactive/streams/operators/spi/ReactiveStreamsFactoryResolver.java

```diff
@@ -77,17 +79,31 @@ private static ReactiveStreamsFactory loadFromSpi(ClassLoader cl) {
         }
 
         if (instance == null) {
-            ServiceLoader<ReactiveStreamsFactory> sl = ServiceLoader.load(
-                    ReactiveStreamsFactory.class, cl);
-            for (ReactiveStreamsFactory spi : sl) {
-                if (instance != null) {
-                    throw new IllegalStateException(
-                            "Multiple ReactiveStreamsFactory implementations found: "
-                                    + spi.getClass().getName() + " and "
-                                    + instance.getClass().getName());
-                } else {
-                    instance = spi;
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
+                    ServiceLoader<ReactiveStreamsFactory> sl = ServiceLoader.load(
+                            ReactiveStreamsFactory.class, cl);
+                    for (ReactiveStreamsFactory spi : sl) {
+                        if (instance != null) {
+                            throw new IllegalStateException(
+                                    "Multiple ReactiveStreamsFactory implementations found: "
+                                            + spi.getClass().getName() + " and "
+                                            + instance.getClass().getName());
+                        } else {
+                            instance = spi;
+                        }
+                    }
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                Throwable t = e.getCause();
```

Comment:
Why does the unwrapping? Why not just throw e?

Reply:
`PrivilegedActionException` is checked so it can't be thrown without wrapping it again.

I guess you could wrap it again in a `RuntimeException` but I think it would be more helpful to the caller to unwrap the `PrivilegedActionException` since that's only created because we called `AccessController.doPrivileged` which is an implementation detail of this method.

Reply:
I agree with @Azquelt - this is what we normally do in the WildFly codebase